src/app/DIAmeterApplication.cpp

```diff
@@ -0,0 +1,558 @@
+#include <cstdio>
+#include "app/tide/abspath.h"
+#include "app/tide/records_to_vector-inl.h"
+
+#include "io/carp.h"
+#include "parameter.h"
+#include "io/SpectrumRecordWriter.h"
+#include "TideIndexApplication.h"
+#include "TideSearchApplication.h"
+#include "DIAmeterApplication.h"
+#include "ParamMedicApplication.h"
+#include "PSMConvertApplication.h"
+#include "tide/mass_constants.h"
+#include "TideMatchSet.h"
+#include "util/Params.h"
+#include "util/FileUtils.h"
+#include "util/StringUtils.h"
+
+
+const double DIAmeterApplication::XCORR_SCALING = 100000000.0;
+const double DIAmeterApplication::RESCALE_FACTOR = 20.0;
+
+DIAmeterApplication::DIAmeterApplication() { /* do nothing */ }
+DIAmeterApplication::~DIAmeterApplication() { /* do nothing */ }
+
+
+int DIAmeterApplication::main(int argc, char** argv) {
+  return main(Params::GetStrings("tide spectra file"));
+}
+int DIAmeterApplication::main(const vector<string>& input_files) {
+  return main(input_files, Params::GetString("tide database"));
+}
+
+int DIAmeterApplication::main(const vector<string>& input_files, const string input_index) {
+  carp(CARP_INFO, "Running tide-search...");
+
+  const string index = input_index;
+  string peptides_file = FileUtils::Join(index, "pepix");
+  string proteins_file = FileUtils::Join(index, "protix");
+  string auxlocs_file = FileUtils::Join(index, "auxlocs");
+
+  // params
+  double bin_width_  = Params::GetDouble("mz-bin-width");
+  double bin_offset_ = Params::GetDouble("mz-bin-offset");
+  vector<int> negative_isotope_errors = TideSearchApplication::getNegativeIsotopeErrors();
+  // bool use_neutral_loss_peaks_ = Params::GetBool("use-neutral-loss-peaks");
+  // bool use_flanking_peaks_ = Params::GetBool("use-flanking-peaks");
+
+  // Read proteins index file
+  ProteinVec proteins;
+  pb::Header protein_header;
+  if (!ReadRecordsToVector<pb::Protein, const pb::Protein>(&proteins, proteins_file, &protein_header)) { carp(CARP_FATAL, "Error reading index (%s)", proteins_file.c_str()); }
+
+  // Read auxlocs index file
+  vector<const pb::AuxLocation*> locations;
+  if (!ReadRecordsToVector<pb::AuxLocation>(&locations, auxlocs_file)) { carp(CARP_FATAL, "Error reading index (%s)", auxlocs_file.c_str()); }
+
+  // Read peptides index file
+  pb::Header peptides_header;
+  HeadedRecordReader* peptide_reader = new HeadedRecordReader(peptides_file, &peptides_header);
+  if ((peptides_header.file_type() != pb::Header::PEPTIDES) || !peptides_header.has_peptides_header()) { carp(CARP_FATAL, "Error reading index (%s)", peptides_file.c_str()); }
+
+  const pb::Header::PeptidesHeader& pepHeader = peptides_header.peptides_header();
+  DECOY_TYPE_T headerDecoyType = (DECOY_TYPE_T)pepHeader.decoys();
+  int decoysPerTarget = pepHeader.has_decoys_per_target() ? pepHeader.decoys_per_target() : 0;
```

Comment:
Do you care about this? Does DIAMeter need multiple decoys per target? If not. Isn't it better to throw an error?